### ISSUE TITLE
automated build option: only on dockerfile change (instead of relying only on git push)

### ISSUE THREAD

**Issue by picoded** [docker/hub-feedback#403]
# TLDR;

Add an additional option to automated builds, where it "triggers only on dockerfile changes". This check for file change will only occur after detecting git push.
# Full Backstory

Currently i have all my various docker files grouped into a (single repository)[https://github.com/picoded/dockerfiles]. 

This is mainly to avoid populating my main project space with a long list of docker git repos, and is made possible as i can setup automated build for a subfolder. A similar approach can be seen in various other users repo.

The issue here however is regarding Automated build. As it blindly triggers on the git push. And as I push a single dockerfile, if automated build is enabled across all my docker hub items. It will be a real burn of resources on docker hub. 

As such I am currently manually triggering my automated builds, which is annoying TBH.

Hence the suggestion, to add an option, to only trigger automated build on dockerfile changes. That way I can re-enable automated build with a clear conscience.
# Implementation type

1) Trigger only when push has changes in the specified docker folder
2) Trigger only when the dockerfile is different from the previous build dockerfile
# Additional Info
- username: picoded

**Comment by jakirkham** [comment]
Personally, based on your code and what you are trying to do, IMHO it is just another case of this issue ( https://github.com/docker/hub-feedback/issues/292 ).

**Comment by jakirkham** [comment]
Also, may want to clean up your link it is pointing to your settings, which I believe is unintended.

**Comment by trinitronx** [comment]
I can still see this being a separate issue when not using multiple `Dockerfile`s.  Example use cases:

Given a GitHub repo project tracking many files unrelated to the built docker image,
When I push a change to files that do not need a new docker image built,
Then Docker Hub should not trigger an Automated Build

Given a GitHub repo project tracking many files unrelated to the built docker image,
When I push a change to the `Dockerfile` OR a file to be included in the docker image to be built,
Then Docker Hub should trigger an Automated Build

**Comment by picoded** [comment]
@jakirkham its more on the triggering of build being skipped, then specifying custom Dockerfile. Im quite happy with the compromise of creating separate folders for each Docker build, and pointing the automated build to those folders.

@trinitronx to support changes of other files, it can be changed to simply trigger the build, only if any files in the relevant folder was changed

**Comment by jakirkham** [comment]
I was only thinking of that in terms of restricting your context to each subdirectory when the `Dockerfile`s are in different locations. However, if your `Dockerfile`s are already in subdirectories, maybe you could restrict the build context for each by specifying the subdirectory when creating the builds? Or are you finding that doesn't behave correctly?

**Comment by picoded** [comment]
@jakirkham: Im currently already the build context to a subdirectory. However when i push changes not relevant to the directory. It triggers a new build. Which I feel, should be optimized out.

An example would be a repo with a dozen over dockerfiles, each in their own directory. And when a single one of them is updated. All of them gets rebuilt.

**Comment by wstrange** [comment]
I have exactly the same issue. I use a consolidated github repo with all my images. A single modification in one triggers a build in all of them.

This issue is costing docker real money in terms of unnecessary builds.

**Comment by wstrange** [comment]
There are times when I might update the contents of a Docker subdir that does not touch the Dockerfile itself. 

Ideally, there would be an option to only rebuild if something in the subdir changed. I think this is possible (git diff?) but probably not trivial to implement.  But again - this is costing docker real money....

**Comment by edevil** [comment]
+1

**Comment by MaxMilton** [comment]
+1

I prefer the single repo approach. Not so great having all my containers rebuilt when making a change to a single build.

**Comment by m-housh** [comment]
+1

**Comment by HariSekhon** [comment]
+1

**Comment by pkennedyr** [comment]
Thanks for feedback. We have prioritized this FR in our backlog and will post an update as soon as the the feature is available.

**Comment by diegoferigo** [comment]
+1

**Comment by koallen** [comment]
+1

**Comment by Jellby** [comment]
+1

**Comment by chaitanyaenr** [comment]
+1

**Comment by grihabor** [comment]
+1

**Comment by GlassOfWhiskey** [comment]
+1

**Comment by grihabor** [comment]
Docker Hub supports building on specific tag push. You can set a regular expression to describe which tags you want to build. 

For example, if you have a Dockerfile to build a ``picoded/python`` image and a Dockerfile to build a ``picoded/ruby`` image, you could use a tag ``python/{commit_hash}`` to trigger python image build and tag ``ruby/{commit_hash}`` to trigger ruby image build.

This solution does not seem great at first, but it has some advantages:
- you can automate tag creation process, so you don't have to build each commit or build by hand
- you don't have create a separate repository for the Dockerfile, because you control which commits are built by Docker Hub
- if your build depends on some context (config file in the directory, etc.), you control when Docker Hub build is triggered, so you can add a git tag when any dependency changes

**Comment by HariSekhon** [comment]
I already build from tags for all the different versions but this really isn't sufficient as I also maintain a master branch for the latest and greatest and then merge improvements from master to tag branches. This works really well in terms of git workflow and automation.

I have started to work around this issue which is really bad as it causes replacement builds with different layer hashes triggering redownloads, plus intermmittent failures for temporarily unavailable upstream repos which get hit too often, and sometimes triggers DockerHub's defences to start failing all builds due to too many logins.

My workaround was to use post build hooks that detect the highest tag and retag it latest and not build off master but really this is a big hack compared to a simpler fix in DockerHub as it needs to be done for several dozen of my projects that populate my https://github.com/HariSekhon/Dockerfiles repo.

It's really a much simply generic fix to just check the Dockerfile's parent directory contents max last changed time before deciding whether to trigger a build if anything changed since the last time.

This is easier to code than it is to discuss so I'm not sure why we're even going round in circles having this conversation tbh.

**Comment by JJ** [comment]
I'm surprised that 5 years after there's still not a good solution for this. Probably the way to go forward would be to create a specific [`hooks/build`](https://docs.docker.com/docker-hub/builds/advanced/) script that does that, overriding the default build script.

**Comment by JJ** [comment]
You can use this; just save it as `hooks/post_checkout` and give it `+x` privs to run:

```perl
#!/usr/bin/env perl

use Git;
my $repo = Git->repository (Directory => '.');

my @modified_files = $repo->command('diff', "--name-only", "HEAD", "HEAD^");

die("No Dockerfile modified in the last commit\n")  unless grep( /Dockerfile/, @modified_files);
```

This will "fail" early (in the post_checkout phase) if there's no Dockerfile modified *in the last commit*, and proceed if there's at least one. It uses [perl](https://dev.to/jj/writing-docker-hub-hooks-in-perl-44h2) and the Git module, which are installed by default in the Docker Hub runner.

**Comment by ZOrfeas** [comment]
+1

**Comment by esromneb** [comment]
I can think of two ways to do this

1) Use branches or tags to trigger when builds happen
```
on:
  workflow_dispatch:
  push:
    branches: [ trigger-build-a ]
```

2) Use the cache, as shown here https://github.com/actions/cache  I haven't tried this but it seems to me you can set a cache entry for a "cookie" file when the build completes. that example shows how to use the hashes of folders to work with the cache so just hash everything that affects that specific docker build, check for the cookie, and then it will only rebuild for commits that have changes to that folder.

**Comment by carlos-a-enriquez** [comment]
Actually, there is a way to "automated builds, where it "triggers only on dockerfile changes"". Checkout the following documentation: 
[Paths mechanism](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore)

As mentioned in the page: "When using the push and pull_request events, you can configure a workflow to run based on what file paths are changed. Path filters are not evaluated for pushes of tags.

Use the paths filter when you want to include file path patterns or when you want to both include and exclude file path patterns. Use the paths-ignore filter when you only want to exclude file path patterns. You cannot use both the paths and paths-ignore filters for the same event in a workflow."
